### ISSUE TITLE
Inconsistent loss between different TP

### ISSUE THREAD

**Issue by weixuansun** [pytorch/torchtitan#1750]
### Bug description

I have encountered different Inconsistent loss between different TP on both llama3 and llama4 moe model.
The toml configs are exactly the same except different tensor parallels.
The seed is set and deterministic is turned on.

tensorboard:
## llama4:
gradnorm:

<img width="1278" height="460" alt="Image" src="https://github.com/user-attachments/assets/f289b37a-b0cf-4de0-aa02-20c0a2af9522" />

loss:
<img width="1266" height="460" alt="Image" src="https://github.com/user-attachments/assets/2b3e1245-5474-4c36-ab26-e2a76588363e" />

## llama3:
gradnorm:
<img width="1278" height="460" alt="Image" src="https://github.com/user-attachments/assets/76a51203-d1f0-46dd-b457-778f2001cb79" />
loss:
<img width="1278" height="460" alt="Image" src="https://github.com/user-attachments/assets/0f42db79-0445-4ea7-b7b8-b361904d0eb7" />

### Versions

toml:

```
# torchtitan Config.toml

[job]
dump_folder = "./outputs"
description = "Llama 3 debug training"
print_args = false
use_for_integration_test = true

[profiling]
enable_profiling = false
save_traces_folder = "profile_trace"
profile_freq = 10
enable_memory_snapshot = false
save_memory_snapshot_folder = "memory_snapshot"

[metrics]
log_freq = 1
disable_color_printing = false
enable_tensorboard = true
save_tb_folder = "tb"
enable_wandb = false

[model]
name = "llama3"
flavor = "debugmodel"
# test folder with tokenizer.json, for debug purpose only
hf_assets_path = "./tests/assets/tokenizer"
# converters = ["float8"]

[optimizer]
name = "AdamW"
lr = 4e-3
eps = 1e-15

[lr_scheduler]
warmup_steps = 2  # lr scheduler warm up, normally 20% of the train steps
decay_ratio = 0.8  # lr scheduler decay ratio, 80% of the train steps
decay_type = "linear"
min_lr_factor = 0.1

[training]
local_batch_size = 1
global_batch_size = 64
seq_len = 2048
max_norm = 1.0  # grad norm clipping
steps = 100000
dataset_type = "hf"  # mmap for megatron style
dataset = "c4_test"  # supported datasets: c4_test (2K), c4 (177M)
dataset_path = "3rdparty/torchtitan/tests/assets/c4_test"
seed = 1234
deterministic = true

[parallelism]
data_parallel_replicate_degree = 1
data_parallel_shard_degree = -1
fsdp_reshard_after_forward = "default" # default / never / always
tensor_parallel_degree = {1 / 4}
enable_async_tensor_parallel = false
pipeline_parallel_degree = 1
pipeline_parallel_schedule = "1F1B"
context_parallel_degree = 1
expert_parallel_degree = 1
expert_tensor_parallel_degree = 1

[checkpoint]
enable_checkpoint = false
folder = "checkpoint"
interval = 10
last_save_model_only = false
export_dtype = "float32"
async_mode = "disabled"  # ["disabled", "async", "async_with_pinned_mem"]

[activation_checkpoint]
mode = "none"  # ["none", "selective", "full"]
selective_ac_option = '2'  # 'int' = ac every positive int layer or 'op', ac based on ops policy

[compile]
enable=false
components = ["model", "loss"]

[float8]
enable_fsdp_float8_all_gather = false
precompute_float8_dynamic_scale_for_fsdp = false
filter_fqns = ["output", "router.gate"]
moe_fqns = ["experts"]
```

torch version: 2.9.0+main.de744ca4b19.post20250818
cuda: 12.4

**Comment by tianyu-l** [comment]
Training loss depends on DP degree which determines global batch size (assuming local batch size is fixed). When you change TP degree, your DP degree changes accordingly.

See https://github.com/pytorch/torchtitan/blob/main/docs/converging.md for details.
Also see https://github.com/pytorch/torchtitan/blob/main/docs/debugging.md#seed-checkpoint-based-reproducibility

**Comment by weixuansun** [comment]
> Training loss depends on DP degree which determines global batch size (assuming local batch size is fixed). When you change TP degree, your DP degree changes accordingly.
> 
> See https://github.com/pytorch/torchtitan/blob/main/docs/converging.md for details. Also see https://github.com/pytorch/torchtitan/blob/main/docs/debugging.md#seed-checkpoint-based-reproducibility

I have fixed global_batch_size between different TP, and can confirm n_tokens_seen respect to step is same across experiments.

**Comment by tianyu-l** [comment]
> I have fixed global_batch_size between different TP, and can confirm n_tokens_seen respect to step is same across experiments.

This is not enough. You have to make sure the parameters are all init with the same value. In torchtitan we can't guarantee this if you don't load the model twice from the same checkpoint.

**Comment by weixuansun** [comment]
> > I have fixed global_batch_size between different TP, and can confirm n_tokens_seen respect to step is same across experiments.
> 
> This is not enough. You have to make sure the parameters are all init with the same value. In torchtitan we can't guarantee this if you don't load the model twice from the same checkpoint.

So that means initial model parameters and loss consistency are not guaranteed between different training runs with different TP?

**Comment by tianyu-l** [comment]
There are two main sources of differences:
1. model param init
2. data loading behavior (keeping global batch size the same is not enough)

For 1, you need to use the same check point, or manually make sure the init are the same
For 2, you need to repeat the DP degree for the torchtitan dataloader to behave the same

In addition, there are other factors:
- different kernel accumulation results
- different communication accumulation results
- etc.

**Comment by weixuansun** [comment]
Thx for the reply. I have conducted a new tests, two models are initialized with the same checkpoint, still, with different TP, the loss is still different:

<img width="902" height="557" alt="Image" src="https://github.com/user-attachments/assets/9fda6be4-a6c2-4344-a147-39aafc2e96a7" />
I am using 8 gpus, local_batch_size = 8, global_batch_size = 64, so two tests mesh settings are:

- ['dp_shard', 'tp'], [2, 4]
- ['dp_shard'], [8]

I am not sure if this loss divergence is expected? it seems that the loss differences are larger than that of kernel/communication related differences?
For 2, that means I need to use more gpus to achieve same DP degree?

**Comment by tianyu-l** [comment]
When I say
> data loading behavior (keeping global batch size the same is not enough)

I mean you have to have the same local batch size and the same dp degree across two runs.

You can do multiple runs with the same dp degree and the same local batch size.
- dp 2
- dp 2 tp 2
- dp 2 tp 4

**Comment by weixuansun** [comment]
> When I say
> 
> > data loading behavior (keeping global batch size the same is not enough)
> 
> I mean you have to have the same local batch size and the same dp degree across two runs.
> 
> You can do multiple runs with the same dp degree and the same local batch size.
> 
> * dp 2
> * dp 2 tp 2
> * dp 2 tp 4

thx, that makes sense, 

- dp 2
- dp 2 tp 4 

have consistent loss

<img width="1232" height="562" alt="Image" src="https://github.com/user-attachments/assets/65819bc1-368f-48ca-8550-39d2209487ce" />

btw, intuitively i think same initialization and same global batch size should have consistent loss, 
it is a common operation to change gpu amount during training, then DP will change and leads to inconsistent loss, 
I wonder could we remove randomness in dataloader with different dp degrees,  is that doable?

**Comment by rakkit** [comment]
>it is a common operation to change gpu amount during training, then DP will change and leads to inconsistent loss,
I wonder could we remove randomness in dataloader with different dp degrees, is that doable?

@weixuansun, we can, but definitely need extra effort.  you can check out the concept of "Heisenbug" and how they fix it by [Mosaic people](https://www.databricks.com/blog/mosaicml-streamingdataset)
